### DISCUSSION TITLE
How to create a function for regexp operator

### DISCUSSION THREAD

**Discussion by nevarman** [linq2db/linq2db#3873] [Q&A] [answered]
Hello everyone,
 
I have been trying to do a simple regex query with the regexp operator. Since regex is not supported with LINQ queries(as I understood from some other discussions) I wanted to use direct queries such as: ``` SELECT * FROM Docs WHERE NAME REGEXP '^.*LOTR.pfd*$'```. 
When I use the direct regexp query I receive an exception that the regexp function is not implemented. I am currently using MS Sqlite and as the documentation here says I wanted to add a function for regex operator: [https://learn.microsoft.com/en-us/dotnet/standard/data/sqlite/user-defined-functions#operators](url).

I have searched the discussions and documentation and could not found a something similar to the ```CreateFunction``` method for the data connection class. I would be glad if someone could lead me to a code snippet or the correct documentation for this method or some similar functionality. 

Thank you.

**Comment by MaceWindu** [comment, marked as answer]
this is method of Microsoft.Data.Sqlite provider. Because custom function registration performed per-connection, best option will be

1. use connection factory, which will configure connection
2. use connection interceptor

---
### Factory

```cs
DbConnection ConnectionFactory()
{
    var cn = new SqliteConnection(connectionString);
    // cn.CreateFunction calls
    return cn;
}
```

and use it:
```cs
// 1. pass factory to DataConnection ctor
new DataConnection(SQLiteTools.GetDataProvider(ProviderName.SQLiteMS), ConnectionFactory);

// 2. use options builder
var options = new LinqToDBConnectionOptionsBuilder().UseConnectionFactory(SQLiteTools.GetDataProvider(ProviderName.SQLiteMS), ConnectionFactory);

var db = DataConnection(options);
// or
var db = DataContext(options);
```

---
### Interceptor

```cs
class MyInterceptor : ConnectionInterceptor
{
    public static readonly ConnectionInterceptor Instance = new MyInterceptor ();

    public override void ConnectionOpening(ConnectionEventData eventData, DbConnection connection)
    {
        CreateFunctions((SqliteConnection)connection);
    }

    public override Task ConnectionOpeningAsync(ConnectionEventData eventData, DbConnection connection, CancellationToken cancellationToken)
    {
        CreateFunctions((SqliteConnection)connection);
        return TaskEx.CompletedTask;
    }

    private void CreateFunctions(SqliteConnection connection)
    {
        // define custom functions
    }
}
```

Use:

```cs
db.AddInterceptor(MyInterceptor.Instance);
```

**Reply by nevarman** [reply]
Perfect, that helped a lot, I used the interceptor method. 
Thank you very much for the fast answer!
